### ISSUE TITLE
Bug: [@typescript-eslint/no-invalid-void-type] false positive when used in a return type where there is a union

### ISSUE THREAD

**Issue by nbilyk** [typescript-eslint/typescript-eslint#8807]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.ts&code=FAFwngDgpgBAqgOwM4FcBGSDGAnAlm2AXhgAoBKGQgPhgDcB7XAE2GCgA8J7sQYAzFAkwhc9BDA4BDALYQANlBIcomFCG4AuUhWrxk6LHgIwAPnUZMyWhsxgBvYDBiYxSXrUlyURCexVruckcYXD5ScGh6MI8vImIAcgEhETF4igcnJxjvIKcAX2A8oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KImjQO0SODABfECqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
type Unsubscribe = () => void

export function example(executor: () => Unsubscribe | void): void {
  const value = executor()
  if (typeof value == 'function') {
    value()
  }
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error"
  }
}
```

### Expected Result

When used as a return type, at least for return types on input parameters, a union including void should be allowed.

It's a common convention to allow for union types in input parameters as a form of overloading. In the provided example, using `Unsubscribe | undefined` instead would not have the desired result.

### Actual Result

The `Unsubscribe | void` return type is marked with a linting error

**Comment by bradzacher** [comment]
This is a subset of https://github.com/typescript-eslint/typescript-eslint/issues/8755
